pydantic/experimental/pipeline.py

```diff
@@ -0,0 +1,654 @@
+"""Experimental pipeline API functionality. Be careful with this API, it's subject to change."""
+
+from __future__ import annotations
+
+import datetime
+import operator
+import re
+from collections import deque
+from collections.abc import Container
+from dataclasses import dataclass
+from decimal import Decimal
+from functools import cached_property, partial
+from typing import TYPE_CHECKING, Any, Callable, Generic, Pattern, Protocol, TypeVar, Union, overload
+
+import annotated_types
+from typing_extensions import Annotated
+
+if TYPE_CHECKING:
+    from pydantic_core import core_schema as cs
+
+    from pydantic import GetCoreSchemaHandler
+
+from pydantic._internal._internal_dataclass import slots_true as _slots_true
+
+__all__ = ['validate_as', 'validate_as_deferred', 'transform']
+
+
+@dataclass(**_slots_true)
+class _ValidateAs:
+    tp: type[Any]
+    strict: bool = False
+
+
+@dataclass
+class _ValidateAsDefer:
+    func: Callable[[], type[Any]]
+
+    @cached_property
+    def tp(self) -> type[Any]:
+        return self.func()
+
+
+@dataclass(**_slots_true)
+class _Transform:
+    func: Callable[[Any], Any]
+
+
+@dataclass(**_slots_true)
+class _PipelineOr:
+    left: _Pipeline[Any, Any]
+    right: _Pipeline[Any, Any]
+
+
+@dataclass(**_slots_true)
+class _PipelineAnd:
+    left: _Pipeline[Any, Any]
+    right: _Pipeline[Any, Any]
+
+
+@dataclass(**_slots_true)
+class _Eq:
+    value: Any
+
+
+@dataclass(**_slots_true)
+class _NotEq:
+    value: Any
+
+
+@dataclass(**_slots_true)
+class _In:
+    values: Container[Any]
+
+
+@dataclass(**_slots_true)
+class _NotIn:
+    values: Container[Any]
+
+
+_ConstraintAnnotation = Union[
+    annotated_types.Le,
+    annotated_types.Ge,
+    annotated_types.Lt,
+    annotated_types.Gt,
+    annotated_types.Len,
+    annotated_types.MultipleOf,
+    annotated_types.Timezone,
+    annotated_types.Interval,
+    annotated_types.Predicate,
+    # common predicates not included in annotated_types
+    _Eq,
+    _NotEq,
+    _In,
+    _NotIn,
+    # regular expressions
+    Pattern[str],
+]
+
+
+@dataclass(**_slots_true)
+class _Constraint:
+    constraint: _ConstraintAnnotation
+
+
+_Step = Union[_ValidateAs, _ValidateAsDefer, _Transform, _PipelineOr, _PipelineAnd, _Constraint]
+
+_InT = TypeVar('_InT')
+_OutT = TypeVar('_OutT')
+_NewOutT = TypeVar('_NewOutT')
+
+
+class _FieldTypeMarker:
+    pass
+
+
+@dataclass(**_slots_true)
+class _Pipeline(Generic[_InT, _OutT]):
+    """Abstract representation of a chain of validation, transformation, and parsing steps."""
+
+    _steps: list[_Step]
```

Comment:
Is there a reason to use a list instead of a tuple here? It just stuck out to me that you use `validate_as = _Pipeline[Any, Any]([]).validate_as` down below, which means the function is actually a method call on a global instance, so I figured it might be a good idea to ensure instances are immutable.

I guess to do that properly you might also want to make the `_Pipeline` dataclass frozen.

I'm assuming there are no APIs that actually mutate a pipeline today but just feels like it would make it clear that that was an assumption. But I'm okay leaving as-is if you prefer.

Reply:
Agreed makes sense might as well make it immutable.

Reply:
Agreed, I just implemented this change! And I learned some things about type hints with tuples and ellipses 🤓, as well as the interactions between frozen dataclasses and generics.

Reply:
New blog post idea - frozen dataclasess and generics 😂